sno/init.py

```diff
@@ -43,200 +57,368 @@ def import_gpkg(ctx, geopackage, table, list_tables):
-        value = os.path.expanduser(value)
-
-        # pass to Click's normal path resolving & validation
-        path = super().convert(value, param, ctx)
-        return (prefix, path, suffix)
+        try:
+            klass = globals()[f'Import{ds.GetDriver().ShortName}']
+        except KeyError:
+            klass = OgrImporter
 
-    def fail(self, message, param=None, ctx=None, exit_code=NO_IMPORT_SOURCE):
-        raise NotFound(message, exit_code=exit_code)
+        return klass(
+            ds, table, writable=writable, source=source, ogr_source=ogr_source,
+        )
 
+    @classmethod
+    def quote_ident_part(cls, part):
+        """
+        SQL92 conformant identifier quoting, for use with OGR-dialect SQL
+        (and most other dialects)
+        """
+        part = part.replace('"', '""')
+        return '"%s"' % part
 
-class ImportGPKG:
-    """ GeoPackage Import Source """
+    @classmethod
+    def quote_ident(cls, *parts):
+        """
+        Quotes an identifier with double-quotes for use in SQL queries.
 
-    prefix = "GPKG"
+            >>> quote_ident('mytable')
+            '"mytable"'
+        """
+        if not parts:
+            raise ValueError("at least one part required")
+        return '.'.join([cls.quote_ident_part(p) for p in parts])
 
-    def __init__(self, source, table=None):
-        self.source = source
+    def __init__(self, ogr_ds, table=None, *, source, ogr_source, writable):
+        self.ds = ogr_ds
+        self.driver = self.ds.GetDriver()
         self.table = table
-        self.db = gpkg.db(self.source)
-
-    def __str__(self):
-        s = f"GeoPackage: {self.source}"
-        if self.table:
-            s += f":{self.table}"
-        return s
-
-    def check(self):
-        dbcur = self.db.cursor()
-
-        sql = "SELECT 1 FROM sqlite_master WHERE type='table' AND name='gpkg_contents';"
-        try:
-            if not dbcur.execute(sql).fetchone():
-                raise ValueError("gpkg_contents table doesn't exist")
-        except (ValueError, apsw.SQLError) as e:
-            raise NotFound(
-                f"'{self.source}' doesn't appear to be a valid GeoPackage",
-                exit_code=NO_IMPORT_SOURCE,
-            ) from e
-
-        if self.table:
-            sql = """
-                SELECT 1
-                FROM gpkg_contents
-                WHERE
-                    table_name=?
-                    AND data_type IN ('features', 'attributes', 'aspatial');"""
-            if not dbcur.execute(sql, (self.table,)).fetchone():
-                raise NotFound(
-                    f"Feature/Attributes table '{self.table}' not found in gpkg_contents",
-                    exit_code=NO_TABLE,
-                )
+        self.source = source
+        self.ogr_source = ogr_source
+        self.writable = writable
 
-    def get_table_list(self):
-        db = gpkg.db(self.source)
-        dbcur = db.cursor()
+    @property
+    @functools.lru_cache(maxsize=1)
```

Comment:
is this safe to do wrt GDAL's GC handling?

Reply:
This just means the layer won't be collected until the OGRImporter is collected (which is at the end of the `import_table` method, just before the process ends). Seems probably-okay to me?

small aside: GDAL's GC handling has always confused me. 
* Why do all the python usage examples do `del ds` at the end of the method, just before it goes out of scope (and thus would be implicitly deleted) _anyway_. (I remove those del statements and have never had any trouble)
* I've noticed before that GDAL's objects get deleted _immediately_ upon going out of scope. They don't hang around waiting for a future GC run. Not quite sure how they implemented this, or what its implications are

Reply:
"they" didn't implement it... IIUC Python's GC deallocates objects immediately when the reference count drops to 0 and the generational GC is only for cleaning up cycles.

Reply:
ok. that makes sense. I'm pretty sure this caching is fine 👍